src/libraries/System.Formats.Asn1/src/System/Formats/Asn1/AsnWriter.cs

```diff
@@ -288,6 +288,43 @@ public TReturn Encode<TState, TReturn>(TState state, Func<TState, ReadOnlySpan<b
                 _encodeDepth--;
             }
         }
+
+        /// <summary>
+        ///   Provides the encoded representation of the data to the specified callback.
+        /// </summary>
+        /// <param name="encodeCallback">
+        ///   The callback that receives the encoded data.
+        /// </param>
+        /// <param name="state">
+        ///   The state to pass to <paramref name="encodeCallback" />.
+        /// </param>
+        /// <typeparam name="TState">
+        ///   The type of the state.
+        /// </typeparam>
+        /// <exception cref="ArgumentNullException">
+        ///   <paramref name="encodeCallback"/> is <see langword="null"/>.
+        /// </exception>
+        /// <exception cref="InvalidOperationException">
+        ///   A <see cref="PushSequence"/> or <see cref="PushSetOf"/> has not been closed via
+        ///   <see cref="PopSequence"/> or <see cref="PopSetOf"/>.
+        /// </exception>
+        public void Encode<TState>(TState state, Action<TState, ReadOnlySpan<byte>> encodeCallback)
+            where TState : allows ref struct
+        {
+            if (encodeCallback is null)
+                throw new ArgumentNullException(nameof(encodeCallback));
+
+            try
+            {
+                _encodeDepth = checked(_encodeDepth + 1);
```

Comment:
If this fails with a checked exception, won't it now erroneously decrement _encodeDepth?

Reply:
...yyyes.